### PR TITLE
Generalized article_and_description

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @matthewjasper 

The logic of finding the right word and article to print seems to be repeated elsewhere... this is an experimental method to unify this logic...

### CODE REVIEW

File: src/librustc_mir/borrow_check/universal_regions.rs [hunk 1]
```diff
@@ -130,6 +130,29 @@ impl<'tcx> DefiningTy<'tcx> {
...
+            _ => false,
+        }
+    }
+
+    pub fn def_id(&self) -> DefId {
+        match *self {
+            DefiningTy::Closure(def_id, ..) => def_id,
+            DefiningTy::Generator(def_id, ..) => def_id,
+            DefiningTy::FnDef(def_id, ..) => def_id,
+            DefiningTy::Const(def_id, ..) => def_id,
```
Comment (Centril) [inline src/librustc_mir/borrow_check/universal_regions.rs]: Use an or-pattern instead.

File: src/librustc_mir/borrow_check/diagnostics/region_errors.rs [hunk 2]
```diff
@@ -622,18 +622,15 @@ impl<'tcx> RegionInferenceContext<'tcx> {
...
+            mbcx.infcx.tcx.article_and_description(self.universal_regions.defining_ty.def_id());
 
         // Revert to the normal error in these cases.
         // Assignments aren't "escapes" in function items.
         if (fr_name_and_span.is_none() && outlived_fr_name_and_span.is_none())
-            || (*category == ConstraintCategory::Assignment && escapes_from == "function")
-            || escapes_from == "const"
+            || (*category == ConstraintCategory::Assignment
+                && self.universal_regions.defining_ty.is_fn_def())
+            || self.universal_regions.defining_ty.is_closure()
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/diagnostics/region_errors.rs]: This should be `is_const`

File: src/librustc/ty/context.rs [hunk 3]
```diff
@@ -1510,6 +1510,20 @@ impl<'tcx> TyCtxt<'tcx> {
...
     }
+
+    /// Returns a displayable description and article for the given `def_id` (e.g. `("a", "closure")`).
+    pub fn article_and_description(
+        &self,
+        def_id: crate::hir::def_id::DefId,
+    ) -> (&'static str, &'static str) {
+        self.def_kind(def_id).map_or_else(
+            || {
+                // TODO: is it a problem to try to use the ty here?
```
Comment (matthewjasper) [inline src/librustc/ty/context.rs]: Looking at what we have available in rustc I would probably start by matching on `tcx.def_key(def_id).disambiguated_data.data` and then using `def_kind` for the `TypeNs`, `ValueNs` and `MacroNs` cases.

Which leaves closures. I would rather not use the type here, since it loses information on async generators. I guess there are a few options here:

* Define a `generator_kind(DefId) -> Option<GeneratorKind>` query.
* Make a method of `MirBorrowckCtxt` that uses `generator_kind` in the MIR body. Make this query only get `closure/generator` correct.

Comment (mark-i-m) [reply src/librustc/ty/context.rs]: > Define a generator_kind(DefId) -> Option<GeneratorKind> query.

@matthewjasper How would I go about doing this?

Comment (matthewjasper) [reply src/librustc/ty/context.rs]: Change the variant of `EntryKind` to be `Generator(hir::GeneratorKind)` (the current contents are unused).
Then copy and modify `static_mutability`.

Comment (mark-i-m) [reply src/librustc/ty/context.rs]: Do you mean this one: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_metadata/rmeta/enum.EntryKind.html#variant.Generator

Comment (mark-i-m) [reply src/librustc/ty/context.rs]: @matthewjasper Could you give a bit more detail? Is there a query to get the `EntryKind` for a def_id?

Comment (matthewjasper) [reply src/librustc/ty/context.rs]: Yes, that's the correct variant.

Queries that use metadata work as follows:
1. Define a version of the query for items in the current crate https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_typeck/collect.rs#L2597-L2607
1. Define a method on `CrateMetadata` (`self.kind` will get the entry): https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_metadata/rmeta/decoder.rs#L1389-L1395
1. Add a provider for external crates that uses that method: https://github.com/rust-lang/rust/blob/6d3f4e0aab3e36ceb8b83d1e9467514685f6b751/src/librustc_metadata/rmeta/decoder/cstore_impl.rs#L136

**Comment by mark-i-m** [comment]
@matthewjasper Thanks! I have pushed my progress so far. Please let me know what you think.

**Comment by mark-i-m** [comment]
@matthewjasper What should I do about `src/librustc_typeck/collect.rs ` which exceeds tidy's maximum number of lines (3004 lines)?

**Comment by matthewjasper** [comment]
~~Wait for #67681~~ 
Pick one of the providers and move it to a submodule. Maybe `codegen_fn_attrs`?

File: src/librustc_metadata/rmeta/mod.rs [hunk 4]
```diff
@@ -301,7 +301,7 @@ enum EntryKind<'tcx> {
     Mod(Lazy<ModData>),
     MacroDef(Lazy<MacroDef>),
     Closure,
-    Generator(Lazy!(GeneratorData<'tcx>)),
+    Generator(Lazy<hir::GeneratorKind>),
```
[the remark this answers is not in the document]
Comment (mark-i-m) [reply src/librustc_metadata/rmeta/mod.rs]: It seems that doing this would require `GeneratorData` to be `Copy`, but that looks like a large struct that probably shouldn't be `Copy`...

[the remark this answers is not in the document]
Comment (matthewjasper) [reply src/librustc_metadata/rmeta/mod.rs]: But the type here is `GeneratorKind`, which is already `Copy`.

File: src/librustc/ty/sty.rs [hunk 5]
```diff
@@ -254,6 +254,41 @@ pub enum TyKind<'tcx> {
     Error,
 }
 
+impl<'tcx> TyKind<'tcx> {
+    pub fn article_and_description(&self) -> (&'static str, &'static str) {
```
Comment (matthewjasper) [inline src/librustc/ty/sty.rs]: This doesn't seem to be used any more, and is somewhat duplicating `sort_string`

File: src/librustc_metadata/rmeta/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mark-i-m) [reply src/librustc_metadata/rmeta/mod.rs]: Actually, yeah, looking back now... I'm not sure what I was thinking...

**Comment by mark-i-m** [comment]
Tests passing now. The UI changes are all "function" -> "method".

File: src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr [hunk 6]
```diff
@@ -7,7 +7,7 @@ LL |     fn foo<'a>(x: &i32, y: &'a i32) -> &'a i32 {
    |            lifetime `'a` defined here
 LL | 
 LL |         if x > y { x } else { y }
-   |                    ^ function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
+   |                    ^ method was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
```
Comment (mark-i-m) [inline src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: Hmm... actually, this is not a method... all the others appear correct.

And I also just realized, that I don't know why this change in UI happened? Any ideas?

Comment (matthewjasper) [reply src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: The change causing this is here: https://github.com/rust-lang/rust/pull/67742/files#diff-5f616dd3631e07b51dd7c59f32ccfb76R433

Comment (mark-i-m) [reply src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: Hmm... do you mean this one instead?

https://github.com/rust-lang/rust/pull/67742/files?file-filters%5B%5D=.rs#diff-5f616dd3631e07b51dd7c59f32ccfb76R505-R507

Also, this seems to imply that we are getting a `DefKind::Method` instead of a `DefKind::Fn`. Is this expected?

Comment (matthewjasper) [reply src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: `DefKind::Method` is for all associated functions. This is generally what `Method` means in the compiler, not that it should be leaking into error messages.

Comment (Centril) [reply src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: We should rename `DefKind::Method` to `DefKind::AssocFn` so that it fits with the scheme used for `AssocTy`, `AssocConst`, etc. cc @petrochenkov @eddyb

**Comment by matthewjasper** [comment]
r=me if you're happy with the current use of "method" in the diagnostics.

**Comment by mark-i-m** [comment]
@matthewjasper Thanks! This looks good to me.

If there is an easy way to fix it, perhaps I can do that in a followup?

File: src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr (hunk 6 above)
Comment (eddyb) [reply src/test/ui/lifetimes/lifetime-errors/ex1-return-one-existing-name-if-else-using-impl.nll.stderr]: Pretty sure we have an issue open about this, yeah.

**Comment by matthewjasper** [comment]
I was thinking of changing what `DefKind::descr` returns, but maybe that should be a separate change because it might change even more diagnostics.

@bors r+

File: src/librustc/ty/context.rs [hunk 7]
```diff
@@ -1512,6 +1512,24 @@ impl<'tcx> TyCtxt<'tcx> {
                 .subst(*self, self.mk_substs([self.lifetimes.re_static.into()].iter())),
         )
     }
+
+    /// Returns a displayable description and article for the given `def_id` (e.g. `("a", "struct")`).
+    pub fn article_and_description(&self, def_id: DefId) -> (&'static str, &'static str) {
+        match self.def_key(def_id).disambiguated_data.data {
+            DefPathData::TypeNs(..) | DefPathData::ValueNs(..) | DefPathData::MacroNs(..) => {
+                let kind = self.def_kind(def_id).unwrap();
+                (kind.article(), kind.descr(def_id))
```
Comment (eddyb) [inline src/librustc/ty/context.rs]: I wish I saw this earlier - I think you should *always* call `def_kind` and only look at the `def_key` or other things if `def_kind` returned `None`.

The reason is that `DefKind` should eventually encompass every possible `DefId`, and the current situation is a transitional one (since `DefKind` was split out of name resolution, which only needs the current set of possible `DefKind`s).

Comment (eddyb) [reply src/librustc/ty/context.rs]: I don't know if there is an issue for finishing `DefKind`, but feel free to.

Ideally `def_key` would be used in as few places as possible, as it's really the "name" of a definition more than anything else.
